src/nbl/asset/IAssetManager.cpp

```diff
@@ -400,4 +423,13 @@ void IAssetManager::insertBuiltinAssets()
         for(auto &path : paths)
             addBuiltInToCaches(pipelineLayout, path);
     }
+
+    /*
+        Pipelines
+    */
+
+    //! Okay, an explaination is needed 
+    {
+
```

Comment:
well, refering to https://github.com/Devsh-Graphics-Programming/Nabla/blob/animation/src/nbl/asset/interchange/CPLYMeshFileLoader.cpp#L674

if I have to builtin the pipeline and insert it into cache I need to know what to do with constructor input parameters, since vertex, blend inputs (etc) are inmutable after pipeline creation. 

and if it comes to https://github.com/Devsh-Graphics-Programming/Nabla/blob/animation/src/nbl/asset/interchange/CPLYMeshFileLoader.cpp#L644

`// this metadata should probably go into pipeline layout's asset bundle (@Crisspl TODO: review)`

okay, API has changed and now I'm able to insert entire metadata upon the bundle being return, so
actually since the descriptor set layout is cached anyways I would move

```
constexpr IRenderpassIndependentPipelineMetadata::E_COMMON_SHADER_INPUT types[DS1_METADATA_ENTRY_CNT] =
		{
			IRenderpassIndependentPipelineMetadata::ECSI_WORLD_VIEW_PROJ,
			IRenderpassIndependentPipelineMetadata::ECSI_WORLD_VIEW,
			IRenderpassIndependentPipelineMetadata::ECSI_WORLD_VIEW_INVERSE_TRANSPOSE
		};
		constexpr uint32_t sizes[DS1_METADATA_ENTRY_CNT] =
		{
			sizeof(SBasicViewParameters::MVP),
			sizeof(SBasicViewParameters::MV),
			sizeof(SBasicViewParameters::NormalMat)
		};
		constexpr uint32_t relOffsets[DS1_METADATA_ENTRY_CNT] =
		{
			offsetof(SBasicViewParameters,MVP),
			offsetof(SBasicViewParameters,MV),
			offsetof(SBasicViewParameters,NormalMat)
		};
		for (uint32_t i = 0u; i < DS1_METADATA_ENTRY_CNT; ++i)
		{
			auto& semantic = (shaderInputsMetadata->end() - i - 1u)[0];
			semantic.type = types[i];
			semantic.descriptorSection.type = IRenderpassIndependentPipelineMetadata::ShaderInput::ET_UNIFORM_BUFFER;
			semantic.descriptorSection.uniformBufferObject.binding = ds1layout->getBindings().begin()[0].binding;
			semantic.descriptorSection.uniformBufferObject.set = 1u;
			semantic.descriptorSection.uniformBufferObject.relByteoffset = relOffsets[i];
			semantic.descriptorSection.uniformBufferObject.bytesize = sizes[i];
			semantic.descriptorSection.shaderAccessFlags = ICPUSpecializedShader::ESS_VERTEX;
		}
```

to **IAssetManager.cpp** and execute `placeMeta`

Reply:
> if I have to builtin the pipeline and insert it into cache I need to know what to do with constructor input parameters, since vertex, blend inputs (etc) are inmutable after pipeline creation.

make several versions, only thing that can change are vertex inputs and there's a limited number of permutations.

PLY doesn't have any blend params that change depending on model.

Reply:
Nothing goes into `IAssetManager.cpp`

if something is a built-in for a specific loader it either gets added to builtins from its constructor... or the `initialize` method

Reply:
> okay, API has changed and now I'm able to insert entire metadata upon the bundle being returned, so
actually since the descriptor set layout is cached anyways I would move

I already moved it
https://github.com/Devsh-Graphics-Programming/Nabla/blob/animation/src/nbl/asset/interchange/IRenderpassIndependentPipelineLoader.cpp

Reply:
hmm, actually I will do the same like glTF does